File: core/claims/family.py
"""Membership in a pairing-friendly family, from a single parameter.

BLS12 and BN curves are not chosen; they are generated. One integer `u`
determines the characteristic, the subgroup order and the trace, through
polynomials fixed by the family. That makes the whole parameter set
checkable from one number, which is worth more here than it sounds.

Three things follow from a family claim that nothing else in the format
supplies for these curves:

**Rigidity.** Prime-field curves in this corpus can say where their
parameters came from — P-256 derives its `b` from a published seed, and
`check.seed-derivation` re-runs it. A pairing curve has no such claim, so
"why these numbers and not others" was answerable only by pointing at a
standard. A family claim answers it arithmetically: these numbers are the
ones the family produces at this `u`, and there was no freedom left to
abuse.

**The embedding degree.** k = 12 is a property of both families rather
than a fact about a particular curve, so a family claim and the
independently computed `curve.embedding` check each other.

**A footing for the twist relation.** Inside a family the twist is fixed
by construction, which is the route out of the gap the format currently
admits.

The polynomials are stated once, here, and used by both the producer and
the tests. The verifier holds its own copy in Rust, which is the point of
having two implementations.
"""

from __future__ import annotations

__all__ = ["FAMILIES", "FamilyError", "derive", "check_against"]


class FamilyError(ValueError):
    """The parameter does not generate the curve it is claimed to."""


def _bls12(u: int) -> tuple[int, int, int]:
    """p, r, t for BLS12 at parameter u.

    r = u^4 - u^2 + 1
    t = u + 1
    p = (u - 1)^2 * r / 3 + u

    The division by three has to be exact. Python's floor division would
    quietly produce a number for a `u` that generates nothing, and the
    result would look like a characteristic. Not every integer is a valid
    BLS12 parameter, and this is where most invalid ones are caught.
    """
    r = u**4 - u**2 + 1
    numerator = (u - 1) ** 2 * r
    quotient, remainder = divmod(numerator, 3)
    if remainder:
        raise FamilyError(
            "(u - 1)^2 * r is not divisible by 3, so u generates no BLS12 curve"
        )
    return quotient + u, r, u + 1


def _bls24(u: int) -> tuple[int, int, int]:
    """p, r, t for BLS24 at parameter u.

    r = u^8 - u^4 + 1
    t = u + 1
    p = (u - 1)^2 * r / 3 + u

    The same shape as BLS12 with the cyclotomic polynomial one step
    further along, and the same exact division by three: a floor division
    would hand back a number for a u that generates nothing, and that
    number would look exactly like a characteristic.
    """
    r = u**8 - u**4 + 1
    numerator = (u - 1) ** 2 * r
    quotient, remainder = divmod(numerator, 3)
    if remainder:
        raise FamilyError(
            "(u - 1)^2 * r is not divisible by 3, so u generates no BLS24 curve"
        )
    return quotient + u, r, u + 1


def _bn(u: int) -> tuple[int, int, int]:
    """p, r, t for BN at parameter u.

    p = 36u^4 + 36u^3 + 24u^2 + 6u + 1
    r = 36u^4 + 36u^3 + 18u^2 + 6u + 1
    t = 6u^2 + 1

    No divisions, so any integer produces a candidate triple; whether it
    is a curve is decided by p and r being prime, which other claims
    establish.
    """
    return (
        36 * u**4 + 36 * u**3 + 24 * u**2 + 6 * u + 1,
        36 * u**4 + 36 * u**3 + 18 * u**2 + 6 * u + 1,
        6 * u**2 + 1,
    )


# Closed on purpose. An unknown name is a refusal rather than a claim
# nobody can check: a verifier that skipped families it did not recognise
# would let a bundle assert membership in anything.
FAMILIES = {
    "bls12": (_bls12, 12),
    "bls24": (_bls24, 24),
    "bn": (_bn, 12),
}


def derive(family: str, u: int) -> tuple[int, int, int]:
    """The characteristic, subgroup order and trace at this parameter."""
    if family not in FAMILIES:
        raise FamilyError(f"unknown family: {family!r}")
    polynomials, _degree = FAMILIES[family]
    p, r, t = polynomials(u)
    if p <= 0 or r <= 0:
        raise FamilyError("u generates a non-positive characteristic or order")
    return p, r, t


def check_against(
    family: str, u: int, p: int, r: int, cardinality: int
) -> tuple[int, int, int]:
    """Derive and compare against numbers other claims already establish.

    The trace is compared as `p + 1 - #E`, in signed arithmetic. Writing
    it as a subtraction of unsigned values would be a trap the day a
    family with a negative trace appears.
    """
    derived_p, derived_r, derived_t = derive(family, u)
    if derived_p != p:
        raise FamilyError("the family polynomial does not give this characteristic")
    if derived_r != r:
        raise FamilyError("the family polynomial does not give this subgroup order")
    if derived_t != p + 1 - cardinality:
        raise FamilyError("the family polynomial does not give this trace")
    return derived_p, derived_r, derived_t

File: verifier/src/family.rs
//! Membership in a pairing-friendly family, re-derived from one integer.
//!
//! BLS12 and BN curves are generated rather than chosen: a single `u`
//! determines the characteristic, the subgroup order and the trace
//! through polynomials the family fixes. Recomputing them is exact
//! integer arithmetic over claims already proved, which is why this sits
//! at the same tier as the other `check.*` kinds.
//!
//! The polynomials are written out here rather than shared with the
//! producer. That is the arrangement the whole project rests on: two
//! implementations, in different languages, that have to agree. A shared
//! constant would let one transcription error pass twice.
//!
//! Everything is computed on signed integers. The trace of a BLS12 curve
//! with negative `u` is itself negative, and `p + 1 - #E` has no business
//! being evaluated in unsigned arithmetic where it would wrap rather than
//! go below zero.

use num_bigint::{BigInt, BigUint};
use num_traits::{One, Signed, Zero};

use crate::claims::{decimal, find_proved_assert};
use crate::json::Node;

/// A signed decimal from the evidence.
///
/// The parameter of a BLS12 curve is usually negative, so this cannot go
/// through `decimal`, which is unsigned by design.
fn signed(text: &str, what: &str) -> Result<BigInt, String> {
    let (negative, digits) = match text.strip_prefix('-') {
        Some(rest) => (true, rest),
        None => (false, text),
    };
    if digits.is_empty() || !digits.bytes().all(|b| b.is_ascii_digit()) {
        return Err(format!("{what}: not a decimal integer"));
    }
    if digits.len() > 1 && digits.starts_with('0') {
        return Err(format!("{what}: leading zero"));
    }
    if negative && digits == "0" {
        return Err(format!("{what}: negative zero is not a spelling of zero"));
    }
    let value = BigInt::parse_bytes(digits.as_bytes(), 10)
        .ok_or_else(|| format!("{what}: not a decimal integer"))?;
    Ok(if negative { -value } else { value })
}

/// p, r and t at this parameter, or a refusal.
fn derive(family: &str, u: &BigInt) -> Result<(BigInt, BigInt, BigInt), String> {
    let one = BigInt::one();
    match family {
        "bls12" => {
            // r = u^4 - u^2 + 1, t = u + 1, p = (u-1)^2 * r / 3 + u
            let u2 = u * u;
            let r = &u2 * &u2 - &u2 + &one;
            let shifted = u - &one;
            let numerator = &shifted * &shifted * &r;
            let three = BigInt::from(3u32);
            // Exact, and checked. A floor division here would produce a
            // number for a u that generates nothing, and the number would
            // look exactly like a characteristic.
            if !(&numerator % &three).is_zero() {
                return Err(
                    "(u - 1)^2 * r is not divisible by 3, so u generates no BLS12 curve"
                        .into(),
                );
            }
            Ok((numerator / three + u, r, u + &one))
        }
        "bls24" => {
            // r = u^8 - u^4 + 1, t = u + 1, p = (u-1)^2 * r / 3 + u.
            // BLS12 one cyclotomic step further along, with the same
            // exactness condition on the division.
            let u2 = u * u;
            let u4 = &u2 * &u2;
            let u8 = &u4 * &u4;
            let r = &u8 - &u4 + &one;
            let shifted = u - &one;
            let numerator = &shifted * &shifted * &r;
            let three = BigInt::from(3u32);
            if !(&numerator % &three).is_zero() {
                return Err(
                    "(u - 1)^2 * r is not divisible by 3, so u generates no BLS24 curve"
                        .into(),
                );
            }
            Ok((numerator / three + u, r, u + &one))
        }
        "bn" => {
            // p = 36u^4 + 36u^3 + 24u^2 + 6u + 1
            // r = 36u^4 + 36u^3 + 18u^2 + 6u + 1
            // t = 6u^2 + 1
            let u2 = u * u;
            let u3 = &u2 * u;
            let u4 = &u2 * &u2;
            let base = BigInt::from(36u32) * &u4 + BigInt::from(36u32) * &u3;
            let linear = BigInt::from(6u32) * u + &one;
            Ok((
                &base + BigInt::from(24u32) * &u2 + &linear,
                &base + BigInt::from(18u32) * &u2 + &linear,
                BigInt::from(6u32) * &u2 + &one,
            ))
        }
        other => Err(format!(
            "unknown family `{other}`; the list is closed, and a family \
             nobody can check is not a claim"
        )),
    }
}

pub fn check_family(
    p: &BigUint,
    asserts: &Node,
    payload: &Node,
    claims: &[Node],
    name: &str,
) -> Result<String, String> {
    let family = payload.str_field("family", name)?;
    if asserts.str_field("family", name)? != family {
        return Err(format!("{name}: the claim and its evidence name different families"));
    }
    let u = signed(payload.str_field("u", name)?, "u")?;
    if signed(asserts.str_field("u", name)?, "u")? != u {
        return Err(format!("{name}: the claim and its evidence give different parameters"));
    }

    let (derived_p, derived_r, derived_t) =
        derive(family, &u).map_err(|e| format!("{name}: {e}"))?;
    if derived_p.is_negative() || derived_r.is_negative() {
        return Err(format!("{name}: u generates a negative characteristic or order"));
    }

    // Against the subject, and against claims that are already proved.
    // Reading them any other way would let a family claim agree with
    // numbers nobody established.
    if derived_p.to_biguint().as_ref() != Some(p) {
        return Err(format!(
            "{name}: the family polynomial does not give the subject's characteristic"
        ));
    }

    let r = decimal(
        find_proved_assert(claims, "curve.order.prime", "n")?,
        "curve.order.prime.n",
    )?;
    if derived_r.to_biguint().as_ref() != Some(&r) {
        return Err(format!(
            "{name}: the family polynomial does not give the proved subgroup order"
        ));
    }

    let cardinality = decimal(
        find_proved_assert(claims, "curve.cardinality", "n")?,
        "curve.cardinality.n",
    )?;
    let trace = BigInt::from(p.clone()) + BigInt::one() - BigInt::from(cardinality);
    if derived_t != trace {
        return Err(format!(
            "{name}: the family polynomial does not give the trace implied by \
             the proved cardinality"
        ));
    }

    Ok(format!(
        "generated by the {family} polynomials at a {}-bit parameter",
        u.magnitude().bits()
    ))
}
